src/closest.c

```diff
@@ -0,0 +1,235 @@
+/* Sebastian Gibb <mail@sebastiangibb.de>
+ */
+
+#include "MsCoreUtils.h"
+
+#include <R.h>
+#include <Rinternals.h>
+#include <math.h>
+
+/**
+ * Find leftmost interval.
+ *
+ * This function is a simplified version of R's findInterval.
+ *
+ * \param x key value to look for.
+ * \param ptable table/haystack where to look for.
+ * \param low lowest/leftmost index to start the search.
+ * \param n highest/rightmost index.
+ * \return index of the leftmost element
+ *
+ * \note For easier implementation we ignore some corner cases here. Index `0`
+ * and `n - 1` have to be handled outside this function. Also if it is called
+ * multiple times (e.g. if x is an array in the calling function) x has to
+ * sorted increasingly (we never lower `low`).
+ *
+ * \sa r-source/src/appl/interv.c
+ */
+R_xlen_t leftmost(double x, double* ptable, R_xlen_t low, R_xlen_t n) {
+    R_xlen_t high = low + 1, mid = high;
+
+    /* same interval as the last one */
+    if (x < ptable[high] && ptable[low] <= x)
+        return low ;
+
+    /* exponentional search */
+    for (R_xlen_t step = 1;; step *= 2) {
+        /* x is still > ptable[high] so we can update low as well to keep search
+         * space small
+         */
+        low = high;
+        high = low + step;
+        if (high >= n || x < ptable[high])
+            break;
+    }
+    high = (high >= n) ? n - 1 : high;
+
+    /* binary search */
+    while(1) {
+        mid = (low + high) / 2;
+        if (mid == low)
+            break;
+        if (x > ptable[mid])
+            low = mid;
+        else
+            high = mid;
+    }
+    return low;
+}
+
+/**
+ * Find closest value to table, keep duplicates.
+ *
+ * \param x key value to look for.
+ * \param table table/haystack where to look for.
+ * \param tolerance allowed tolerance to be accepted as match, has to be of
+ * length == length(tolerance).
+ * \param nomatch value that should be returned if a key couldn't be matched.
+ * \return index the closest element
+ *
+ * \note x and table have to be sorted increasingly and not containing any NA.
+ */
+SEXP C_closest_dup_keep(SEXP x, SEXP table, SEXP tolerance, SEXP nomatch) {
+    R_xlen_t nx = XLENGTH(x);
+    double* px = REAL(x);
+
+    R_xlen_t ntable = XLENGTH(table);
+    double* ptable = REAL(table);
+
+    double* ptolerance = REAL(tolerance);
+
+    SEXP out = PROTECT(allocVector(REALSXP, nx));
```

Comment:
I did not quite undestand the benefit of using here `REALSXP` instead of `INTSXP` - I would have expected the index to be returned as an integer not as a real.

Reply:
This may be over-engineered but this is needed to support *long vectors*:
(from `?"long vectors"`, highlighting by myself)
> Prior to R 3.0.0, all vectors in R were restricted to at most **2^31 - 1
> elements** and could be **indexed by integer** vectors.

> Currently all atomic (raw, logical, integer, numeric, complex,
> character) vectors, lists and expressions can be much longer on
> 64-bit platforms: such vectors are referred to as ‘long vectors’
> and have a slightly different internal structure. In theory they
> can contain up to **2^52 elements**, but address space limits of
> current CPUs and OSes will be much smaller.  Such objects will
> have a length that is expressed as a double, and can be **indexed by
> double** vectors.

So the question is: Is supporting *long vectors* (spectra with more than `2^31 - 1` elements) needed/useful at all.

(Just as anecdote: in `MALDIquant` I had `integer` for intensity values but were hit by integer overflows with my second data set (that contained way more peaks) for most operations (like TIC, noise estimation, ...). So I had to switch to `double` for intensity values. But currently I can't really imagine to hit the `2^31` limit with a single spectrum, but maybe with a library?)

Reply:
hm, that's a valid question. I can not quite imagine having a spectrum with that many data points - at least not for LC-MS applications. Also, libraries (those that we use for metabolomics) usually have *clean* spectra with only few peaks in it - if you have 100 it's already a lot.

there's also the higher memory demand of double over integer - but that should not be a real problem here. So, I'm fine with either solution but somehow would prefer integer - but you can also keep `double`, but then you should update the help page saying that `closest` returns a `numeric` and not an `integer`.

Reply:
Will change to `integer` and drop long vector support.